tests/unit/test_sysctl.py

```diff
@@ -0,0 +1,332 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import unittest
+from pathlib import Path
+from subprocess import CalledProcessError
+from unittest.mock import call, mock_open, patch
+
+from charms.operator_libs_linux.v0 import sysctl
+
+permission_failure_output = """sysctl: permission denied on key "vm.swappiness", ignoring"""
+partial_permission_failure_output = """sysctl: permission denied on key "vm.swappiness", ignoring
+net.ipv4.tcp_max_syn_backlog = 4096
+"""
+
+TEST_OTHER_CHARM_FILE = """vm.swappiness=60
+net.ipv4.tcp_max_syn_backlog=4096
+"""
+TEST_MERGED_FILE = """# This config file was produced by sysctl lib v0.1
+#
+# This file represents the output of the sysctl lib, which can combine multiple
+# configurations into a single file like.
+vm.max_map_count = 262144
+vm.swappiness=0
+
+"""
+TEST_MULTIPLE_MERGED_FILE = [
+    "# This config file was produced by sysctl lib v0.2\n#\n# This file represents the output of the sysctl lib, which can combine multiple\n# configurations into a single file like.\n",
+    "vm.swappiness=60\n",
+    "net.ipv4.tcp_max_syn_backlog=4096\n",
+]
+TEST_UPDATE_MERGED_FILE = [
+    "# This config file was produced by sysctl lib v0.2\n#\n# This file represents the output of the sysctl lib, which can combine multiple\n# configurations into a single file like.\n",
+    "vm.max_map_count=25500\n",
+]
+
+
+def check_output_side_effects(*args, **kwargs):
+    if args[0] == ["sysctl", "vm.swappiness", "-n"]:
+        return "1"
+    elif args[0] == ["sysctl", "other_value", "-n"]:
+        return "5"
+    elif args[0] == ["sysctl", "vm.swappiness=1", "other_value=5"]:
+        return "1\n5"
+    elif args[0] == ["sysctl", "vm.swappiness=0"]:
+        return permission_failure_output
+    elif args[0] == ["sysctl", "vm.swappiness=0", "net.ipv4.tcp_max_syn_backlog=4096"]:
+        return partial_permission_failure_output
+    elif args[0] == ["sysctl", "exception"]:
+        raise CalledProcessError(returncode=1, cmd=args[0], output="error on command")
+
+    # Tests on 'update()'
+    elif args[0] == ["sysctl", "vm.max_map_count", "-n"]:
+        return "25000"
+    elif args[0] == ["sysctl", "vm.max_map_count=25500"]:
+        return "25500"
+
+
+class TestSysctlConfig(unittest.TestCase):
+    def setUp(self) -> None:
+        self.desired_values = {"vm.swappiness": {"value": 0}}
+        self.loaded_values = {"vm.swappiness": "60", "vm.max_map_count": "25500"}
+
+    @patch("pathlib.Path.exists")
+    @patch("pathlib.Path.glob")
+    @patch("builtins.open", new_callable=mock_open, read_data="vm.max_map_count=25500\n")
+    @patch("charms.operator_libs_linux.v0.sysctl.check_output")
+    @patch("charms.operator_libs_linux.v0.sysctl.Config._create_charm_file")
+    @patch("charms.operator_libs_linux.v0.sysctl.Config._load_data")
```

Comment:
Yeah, this is a lot of mocking... I think we'd be better to just change the directory and use real file operations. We're testing a lot more that way. We can have one test with mocked file operations if you want, to ensure that we're actually writing to the right path if the path isn't overridden.

Reply:
It is a lot, but I was trying to avoid creating files here. This happens more on the `update` call which is the one that holds most of the logic. Would adding integration tests (and leaving unit tests like this) cover your concern here?

Reply:
I'm not against so many `@patch`, but I still think that we should change `SYSCTL_DIRECTORY` to some tmp path to be sure that we will not change anything in system sysctl config and maybe also mocked all subprocess calls so unit tests will not call `sysctl key=value`. I meant in `setUp` function.